osmnx/pois.py

```diff
@@ -37,36 +38,116 @@ def parse_poi_query(north, south, east, west, amenities=None, timeout=180, maxsi
-
-    return query_str
-
-
-def osm_poi_download(polygon=None, amenities=None, north=None, south=None, east=None, west=None,
-                     timeout=180, max_query_area_size=50*1000*50*1000):
+    # Convert input into dict of tags
+    _tags = {}
+    if amenities is not None:
+        # Instatiate amenity key
+        _tags['amenity'] = _tags.get('amenity', [])
+        # Add amenities to _tags dict
+        [_tags['amenity'].append(a) for a in amenities]
+
+    if tags is not None:
+        for k, v in tags.items():
+            if type(v) == bool:
+                if v:
+                    # Use None to signal later that the query should be
+                    # ["tag_key"]
+                    # which gets all tags with key `tag_key`
+                    _tags[k] = None
+
+                continue
+
+            if type(v) == list:
+                # Instatiate key
+                _tags[k] = _tags.get(k, [])
+                [_tags[k].append(item) for item in v]
+
+                continue
+
+            if type(v) == str:
+                # Instatiate key
+                _tags[k] = _tags.get(k, [])
+                _tags[k].append(v)
+
+    # List of dictionaries with one key and one value.
+    # Overpass QL doesn't have an OR, so you can't do "amenity"="a|b"
+    # Instead you have to do "amenity"="a" and then also "amenity"="b"
+    _tags_list = []
+    for k, v in _tags.items():
+        if v is None:
+            _tags_list.append({k: None})
+            continue
+
+        for item in v:
+            _tags_list.append({k: item})
+
+    # Create query template
+    # Build query for each tag for each of nodes, ways, and relations.
+
+    # Set query settings: Query settings must all be on the same line. The body
+    # of the query can be over separate lines. See documentation here:
+    # https://wiki.openstreetmap.org/wiki/Overpass_API/Overpass_QL#Settings
+    query_settings = [
+        '[out:json]',
+        '[timeout:{t}]'.format(t=timeout),
+        '[bbox:{s:.6f},{w:.6f},{n:.6f},{e:.6f}]'.format(s=south, w=west,
+                                                        n=north, e=east)
+    ]
+    if maxsize is not None:
+        query_settings.append('[maxsize:{m}]'.format(m=maxsize))
+
+    query = [
+        ''.join(query_settings) + ';',
+        '('
+    ]
+
+    for d in _tags_list:
+        for k, v in d.items():
+            tag_str = '["{k}"'.format(k=k)
+            if v is not None:
+                tag_str += '="{v}"'.format(v=v)
+            # The `(._; >;);` joins the new result set into the previous result
+            # set: https://wiki.openstreetmap.org/wiki/Overpass_API/Overpass_QL#Item
+            tag_str += '];(._;>;);'
+
+            for t in ['node', 'way', 'relation']:
+                query.append('({t}{tag_str});'.format(t=t, tag_str=tag_str))
+
+    query.append(');')
+    query.append('out;')
+    return ''.join(query)
```

Comment:
One thing that may be useful: could you provide an example returned value as a commented demonstration? What does the resulting string look like at the end?